EIPS/eip-6551.md

```diff
@@ -0,0 +1,393 @@
+
+This proposal grants every ERC-721 token the full capabilities of an Ethereum account while maintaining backwards compatibility with previously deployed ERC-721 token contracts. It does so by deploying unique, deterministically-addressed smart contract accounts for each ERC-721 token via a permissionless registry.
+
+Each token bound account is owned by a single ERC-721 token, allowing the token to interact with the blockchain, record transaction history, and own on-chain assets. Control of each token bound account is delegated to the owner of the ERC-721 token, allowing the owner to initiate on-chain actions on behalf of their token.
+
+Token bound accounts are compatible out of the box with nearly all existing infrastructure that supports Ethereum accounts, from on-chain protocols to off-chain indexers. Token bound accounts can own any type of on-chain asset, and can be extended to support new asset types created in the future.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+### Overview
+
+The system outlined in this proposal has two main components:
+
+- A permissionless registry for deploying token bound accounts
+- A standard interface for token bound account implementations
+
+The following diagram illustrates the relationship between ERC-721 tokens, ERC-721 token owners, token bound accounts, and the Registry:
+![](../assets/eip-6551/diagram.png)
+
+### Registry
+
+The registry serves as a single entry point for projects wishing to utilize token bound accounts. It has two functions:
+
+- `createAccount` - deploys a token bound account for an ERC-721 token given an `implementation` address
+- `account` - a read-only function that computes the token bound account address for an ERC-721 token given an `implementation` address
+
+The registry SHALL deploy each token bound account as an [ERC-1167](./eip-1167.md) minimal proxy with immutable arguments.
+
+The the deployed bytecode of each token bound account SHALL have the following structure:
+
+```
+ERC-1167 Header               (10 bytes)
+<implementation (address)>    (20 bytes)
+ERC-1167 Footer               (15 bytes)
+STOP code                     (1 byte)
+<chainId (uint256)>           (32 bytes)
+<tokenContract (address)>     (32 bytes)
+<tokenId (uint256)>           (32 bytes)
+```
+
+For example, the token bound account with implementation address `0xbebebebebebebebebebebebebebebebebebebebe`, chain ID `1`, token contract `0xcfcfcfcfcfcfcfcfcfcfcfcfcfcfcfcfcfcfcfcf` and token ID `123` would have the following deployed bytecode:
+
+```!
```

Comment:
I'm not familiar with the `!` notation. What is it supposed to do?

Reply:
Ooops, composed this in hackmd where the `!` turns on line wrapping. Will remove :)